exec/exec.go

```diff
@@ -0,0 +1,99 @@
+package exec
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os/exec"
+	"time"
+)
+
+// dcos-go/exec is a os/exec wrapper. It implements io.Reader and can be used to read both STDOUT and STDERR.
+//
+// Usage:
+// ce := exec.Run("bash", []string{"infinite.sh"}, time.Second*3)
+//
+// io.Copy(os.Stdout, ce)
+// err := <- ce.Done
+// if err != nil {
+// 	log.Error(err)
+// }
+
+var (
+	// ErrContextRequired is the error returned by `func run` if the context is empty.
+	ErrContextRequired = errors.New("context cannot be empty")
+)
+
+// CommandExecutor is a structure returned by exec.Run
+// Cancel can be used by a user to interrupt a command execution.
+// Done is a channel the user can read in order to retrieve execution status. Possible statuses:
+//  <nil> command executed successfully, returned 0 exit code
+//  <exit status N> where N is non 0 exit status.
+//  <context deadline exceeded> means timeout was reached and command was killed.
+//  <context canceled>  means that command was canceled by a user.
+type CommandExecutor struct {
+	Cancel context.CancelFunc
+	Done   chan error
+
+	done chan error
+	pipe *io.PipeReader
+}
+
+// Read implements the io.Reader.
+// CommandExecutor will read from stdout and stderr
+func (c CommandExecutor) Read(p []byte) (int, error) {
+	return c.pipe.Read(p)
+}
+
+// RunWithTimeout is a main command in a package. It executes the command passed by a user and returns `CommandExecutor`
+// structure. User is able to cancel the command by running `CommandExecutor.Cancel()` function, wait for
+// `timeout` or wait for command to execute. The error will be sent to CommandExecutor.Done channel
+func RunWithTimeout(command string, arg []string, timeout time.Duration) (CommandExecutor, error) {
+	if timeout <= 0 {
+		return CommandExecutor{}, errors.New("Timeout cannot be negative or empty")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	commandExecutor := CommandExecutor{
+		Cancel: cancel,
+		Done:   make(chan error, 1),
+		done:   make(chan error, 1),
+	}
+
+	return run(ctx, &commandExecutor, command, arg)
+}
+
+func run(ctx context.Context, cmdExec *CommandExecutor, command string, arg []string) (CommandExecutor, error) {
+	if ctx == nil {
+		return *cmdExec, ErrContextRequired
```

Comment:
this looks weird. I'd expect to see `return CommandExecutor{}, ErrContextRequired`